File: 12-unions/main.go
package main

import (
	"fmt"
	"math"
)

// === Go 没有联合体，用接口模拟 tagged union ===

// Shape 接口模拟 tagged union
type Shape interface {
	Area() float64
	Name() string
}

type Circle struct {
	Radius float64
}

func (c Circle) Area() float64 { return math.Pi * c.Radius * c.Radius }
func (c Circle) Name() string  { return "Circle" }

type Rectangle struct {
	Width, Height float64
}

func (r Rectangle) Area() float64 { return r.Width * r.Height }
func (r Rectangle) Name() string  { return "Rectangle" }

type Triangle struct {
	Base, Height float64
}

func (t Triangle) Area() float64 { return 0.5 * t.Base * t.Height }
func (t Triangle) Name() string  { return "Triangle" }

// === 另一种模拟方式：struct + type tag ===

type TokenType int

const (
	TokenNumber TokenType = iota
	TokenString
	TokenPlus
	TokenMinus
	TokenEOF
)

type Token struct {
	Type     TokenType
	NumVal   float64 // 仅 TokenNumber 时有意义
	StrVal   string  // 仅 TokenString 时有意义
}

func (t Token) String() string {
	switch t.Type {
	case TokenNumber:
		return fmt.Sprintf("Number(%g)", t.NumVal)
	case TokenString:
		return fmt.Sprintf("String(%q)", t.StrVal)
	case TokenPlus:
		return "Plus"
	case TokenMinus:
		return "Minus"
	case TokenEOF:
		return "EOF"
	default:
		return "Unknown"
	}
}

// === 用指针模拟可选值 ===
func findIndex(data string, target byte) *int {
	for i := 0; i < len(data); i++ {
		if data[i] == target {
			idx := i
			return &idx
		}
	}
	return nil
}

// === 用多返回值模拟 Result ===
func parseDigit(c byte) (uint8, error) {
	if c >= '0' && c <= '9' {
		return c - '0', nil
	}
	return 0, fmt.Errorf("invalid character: %c", c)
}

func main() {
	// === 接口模拟 tagged union ===
	shapes := []Shape{
		Circle{Radius: 5.0},
		Rectangle{Width: 4.0, Height: 6.0},
		Triangle{Base: 3.0, Height: 8.0},
	}

	fmt.Println("=== 形状面积 ===")
	for _, shape := range shapes {
		fmt.Printf("  %s: %.2f\n", shape.Name(), shape.Area())
	}

	// === 类型断言 switch（类似 match） ===
	fmt.Println("\n=== 类型断言 ===")
	for _, shape := range shapes {
		switch s := shape.(type) {
		case Circle:
			fmt.Printf("  圆的半径: %g\n", s.Radius)
		case Rectangle:
			fmt.Printf("  矩形: %gx%g\n", s.Width, s.Height)
		case Triangle:
			fmt.Printf("  三角形: base=%g, h=%g\n", s.Base, s.Height)
		}
		// 注意：没有穷尽性检查！如果添加新类型但忘记处理，编译器不会报错
	}

	// === struct + tag 模拟 ===
	tokens := []Token{
		{Type: TokenNumber, NumVal: 42},
		{Type: TokenPlus},
		{Type: TokenNumber, NumVal: 8},
		{Type: TokenEOF},
	}

	fmt.Println("\n=== 词法标记 ===")
	for _, token := range tokens {
		fmt.Printf("  %s\n", token)
	}

	// === nil 作为可选值 ===
	fmt.Println("\n=== 可选值（nil） ===")
	data := "Hello, World!"
	if idx := findIndex(data, 'W'); idx != nil {
		fmt.Printf("找到 'W' 在位置: %d\n", *idx)
	} else {
		fmt.Println("未找到")
	}

	if idx := findIndex(data, 'Z'); idx != nil {
		fmt.Printf("找到 'Z' 在位置: %d\n", *idx)
	} else {
		fmt.Println("未找到 'Z'")
	}

	// === 多返回值模拟 Result ===
	fmt.Println("\n=== 错误处理 ===")
	if digit, err := parseDigit('5'); err == nil {
		fmt.Printf("解析 '5': %d\n", digit)
	} else {
		fmt.Printf("错误: %v\n", err)
	}

	if digit, err := parseDigit('x'); err == nil {
		fmt.Printf("解析 'x': %d\n", digit)
	} else {
		fmt.Printf("解析 'x' 错误: %v\n", err)
	}

	// === 接口的局限性 ===
	fmt.Println("\n=== Go 模拟 tagged union 的问题 ===")
	fmt.Println("1. 没有穷尽性检查：添加新类型时编译器不提醒")
	fmt.Println("2. 接口可以被任意类型实现：无法限制变体集合")
	fmt.Println("3. struct+tag 方式浪费内存：每个 token 都存所有字段")
	fmt.Println("4. struct+tag 方式无类型安全：可以访问错误的字段")
}

File: 12-unions/main.rs
use std::f64::consts::PI;

// === Rust 的 enum 就是 tagged union ===
enum Shape {
    Circle(f64),
    Rectangle { width: f64, height: f64 },
    Triangle { base: f64, height: f64 },
}

impl Shape {
    fn area(&self) -> f64 {
        match self {
            Shape::Circle(r) => PI * r * r,
            Shape::Rectangle { width, height } => width * height,
            Shape::Triangle { base, height } => 0.5 * base * height,
        }
    }

    fn name(&self) -> &str {
        match self {
            Shape::Circle(_) => "Circle",
            Shape::Rectangle { .. } => "Rectangle",
            Shape::Triangle { .. } => "Triangle",
        }
    }
}

// === Token：更复杂的 tagged union ===
#[derive(Debug)]
enum Token {
    Number(f64),
    Str(String),
    Plus,
    Minus,
    Eof,
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Token::Number(n) => write!(f, "Number({})", n),
            Token::Str(s) => write!(f, "String(\"{}\")", s),
            Token::Plus => write!(f, "Plus"),
            Token::Minus => write!(f, "Minus"),
            Token::Eof => write!(f, "EOF"),
        }
    }
}

// === Option<T> 和 Result<T, E> 是标准库的 tagged union ===
fn find_index(data: &str, target: char) -> Option<usize> {
    data.find(target)
}

#[derive(Debug)]
enum ParseError {
    InvalidCharacter(char),
}

impl std::fmt::Display for ParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            ParseError::InvalidCharacter(c) => write!(f, "invalid character: {}", c),
        }
    }
}

fn parse_digit(c: char) -> Result<u8, ParseError> {
    if c.is_ascii_digit() {
        Ok(c as u8 - b'0')
    } else {
        Err(ParseError::InvalidCharacter(c))
    }
}

// === 递归 tagged union（链表） ===
#[derive(Debug)]
enum List {
    Cons(i32, Box<List>),
    Nil,
}

impl List {
    fn new() -> Self {
        List::Nil
    }

    fn push(self, val: i32) -> Self {
        List::Cons(val, Box::new(self))
    }

    fn to_vec(&self) -> Vec<i32> {
        let mut result = Vec::new();
        let mut current = self;
        while let List::Cons(val, next) = current {
            result.push(*val);
            current = next;
        }
        result
    }
}

fn main() {
    // === Tagged union 基本使用 ===
    let shapes = vec![
        Shape::Circle(5.0),
        Shape::Rectangle { width: 4.0, height: 6.0 },
        Shape::Triangle { base: 3.0, height: 8.0 },
    ];

    println!("=== 形状面积 ===");
    for shape in &shapes {
        println!("  {}: {:.2}", shape.name(), shape.area());
    }

    // === match 解构（必须穷尽） ===
    println!("\n=== match 解构 ===");
    for shape in &shapes {
        match shape {
            Shape::Circle(r) => println!("  圆的半径: {}", r),
            Shape::Rectangle { width, height } => println!("  矩形: {}x{}", width, height),
            Shape::Triangle { base, height } => println!("  三角形: base={}, h={}", base, height),
        }
    }

    // === Token 示例 ===
    let tokens = vec![
        Token::Number(42.0),
        Token::Plus,
        Token::Number(8.0),
        Token::Eof,
    ];

    println!("\n=== 词法标记 ===");
    for token in &tokens {
        println!("  {}", token);
    }

    // === Option<T> ===
    println!("\n=== Option<T> ===");
    let data = "Hello, World!";

    if let Some(index) = find_index(data, 'W') {
        println!("找到 'W' 在位置: {}", index);
    }

    if let Some(index) = find_index(data, 'Z') {
        println!("找到 'Z' 在位置: {}", index);
    } else {
        println!("未找到 'Z'");
    }

    // unwrap_or 提供默认值（类似 Zig 的 orelse）
    let idx = find_index(data, 'Z').unwrap_or(0);
    println!("unwrap_or 默认值: {}", idx);

    // === Result<T, E> ===
    println!("\n=== Result<T, E> ===");
    match parse_digit('5') {
        Ok(digit) => println!("解析 '5': {}", digit),
        Err(e) => println!("错误: {}", e),
    }

    match parse_digit('x') {
        Ok(digit) => println!("解析 'x': {}", digit),
        Err(e) => println!("解析 'x' 错误: {}", e),
    }

    // ? 操作符（类似 Zig 的 try）在返回 Result 的函数中使用

    // === 递归 tagged union ===
    println!("\n=== 递归 tagged union（链表） ===");
    let list = List::new().push(3).push(2).push(1);
    println!("链表: {:?}", list.to_vec());

    // === 内存大小 ===
    println!("\n=== 内存大小 ===");
    println!("Shape 大小: {} 字节", std::mem::size_of::<Shape>());
    println!("Token 大小: {} 字节", std::mem::size_of::<Token>());
    println!("Option<usize> 大小: {} 字节", std::mem::size_of::<Option<usize>>());
    println!("Result<u8, ParseError> 大小: {} 字节", std::mem::size_of::<Result<u8, ParseError>>());
}
